Assets/HoloToolkit/UX/Scripts/Lines/Spline.cs

```diff
@@ -0,0 +1,423 @@
+                        } else if (pointIndex == NumPoints) {
+                            points[0].Point = point;
+                            points[1].Point += delta;
+                            points[pointIndex - 1].Point += delta;
+                        } else {
+                            points[pointIndex - 1].Point += delta;
+                            points[pointIndex + 1].Point += delta;
+                        }
+                    } else {
+                        if (pointIndex > 0) {
+                            points[pointIndex - 1].Point += delta;
+                        }
+                        if (pointIndex + 1 < points.Length) {
+                            points[pointIndex + 1].Point += delta;
+                        }
+                    }
+                }
+            }
+
+            points[pointIndex].Point = point;
+
+            ForceUpdateAlignment(pointIndex);
+        }
+
+        protected override Vector3 GetUpVectorInternal(float normalizedLength) {
+
+            float arrayValueLength = 1f / points.Length;
+            int indexA = Mathf.FloorToInt(normalizedLength * points.Length);
+            if (indexA >= points.Length)
+                indexA = 0;
+
+            int indexB = indexA + 1;
+            if (indexB >= points.Length)
+                indexB = 0;
+
+            float blendAmount = (normalizedLength - (arrayValueLength * indexA)) / arrayValueLength;
+            Quaternion rotation = Quaternion.Lerp(points[indexA].Rotation, points[indexB].Rotation, blendAmount);
+            return rotation * transform.up;
+        }
+
+        protected override float GetUnclampedWorldLengthInternal()
+        {
+            // Crude approximation
+            // TODO optimize
+            float distance = 0f;
+            Vector3 last = GetPoint(0f);
+            for (int i = 1; i < 10; i++)
+            {
+                Vector3 current = GetPoint((float)i / 10);
+                distance += Vector3.Distance(last, current);
+            }
+            return distance;
+        }
+
+#if UNITY_EDITOR
+        [UnityEditor.CustomEditor(typeof(Spline))]
+        public class CustomEditor : LineBaseEditor
+        {
+            private static bool editPositions = true;
+            private static bool editRotations = false;
+
+            private const float overlappingPointThreshold = 0.015f;
+
+            // Convenience buttons for adding / removing points
+            protected override void DrawCustomFooter()
+            {
+
+                base.DrawCustomFooter();
+
+                Spline line = (Spline)target;
+                
+                HashSet<int> overlappingPointIndexes = new HashSet<int>();
+
+                if (DrawSectionStart(line.name + " Points", "Point Editing"))
+                {
+                    if (GUILayout.Button(" + Add Points to Start"))
+                    {
+                        List<SplinePoint> points = new List<SplinePoint>();
```

Comment:
We can probably cache this list and clear it each time we add a point

Reply:
Cached as a static variable.